packages/dds/tree/src/rebase/rebaser.ts

```diff
@@ -0,0 +1,139 @@
+ */
+
+import { brand, Brand, Contravariant, Covariant, Invariant } from "../util";
+import { UpPath } from "./pathTree";
+
+/**
+ * A way to refer to a particular revision within a given {@link Rebaser} instance.
+ */
+export type RevisionTag = Brand<number, "rebaser.RevisionTag">;
+
+/**
+ * A collection of branches which can rebase changes between them.
+ */
+export class Rebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> {
+    private lastRevision = 0;
+
+    private makeRevision(): RevisionTag {
+        this.lastRevision++;
+        return brand(this.lastRevision);
+    }
+
+    public readonly empty: RevisionTag = brand(0);
+
+    /**
+     * All the actual state needed to do the rebases.
+     *
+     * Source and destination can both walk this to find common ancestor,
+     * then rebase across using changes found on walk.
+     */
+    private readonly branches: Map<RevisionTag, { before: RevisionTag; change: SetFromChangeRebaser<TChangeRebaser>; }>
+        = new Map();
+
+    public constructor(public readonly rebaser: TChangeRebaser) {
+        // TODO
+    }
+
+    /**
+     * Rebase `changes` from being applied to the `from` state to able to be applied to the `to` state.
+     * @returns a RevisionTag for the state after applying changes to `to`, and the rebased changes themselves.
+     */
+    public rebase(changes: ChangeFromChangeRebaser<TChangeRebaser>, from: RevisionTag, to: RevisionTag):
+        [RevisionTag, FinalFromChangeRebaser<TChangeRebaser>] {
+        const initalChangeset: SetFromChangeRebaser<TChangeRebaser> = this.rebaser.import(changes);
+        if (from !== to) {
+            throw Error("Not implemented"); // TODO: rebase
+        }
+        const over = this.rebaser.compose([]);
+        const finalChangeset: SetFromChangeRebaser<TChangeRebaser> = this.rebaser.rebase(initalChangeset, over);
+        const newRevision = this.makeRevision();
+        this.branches.set(newRevision, { before: to, change: finalChangeset });
+        const output: FinalFromChangeRebaser<TChangeRebaser> = this.rebaser.export(finalChangeset);
+        return [newRevision, output];
+    }
+
+    /**
+     * Informs the Rebaser that `revision` will not be used again,
+     * and internal resources related to it may be freed.
+     */
+    public discardRevision(revision: RevisionTag): void {
+        throw Error("Not implemented"); // TODO
+    }
+}
+
+// TODO: managing the types with these is not working well (inferring any for methods in Rebaser). Do something else.
+
+export type ChangeFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
+    TChangeRebaser extends ChangeRebaser<infer TChange, any, any> ? TChange : never;
+
+export type FinalFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
+    TChangeRebaser extends ChangeRebaser<any, infer TFinal, any> ? TFinal : never;
+
+export type SetFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
```

Comment:
```suggestion
export type StateFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
```
You'll need to update usages as well.

Reply:
It was the name of the identifier in the infer block that was wrong here.